### PR TITLE
Add surveys

### LINKED ISSUE #13: Surveys
#### :tophat: User Story

TLDR: We need to make an online version of this: https://drive.google.com/file/d/0B2voJ3vHaOfJY0tCUFlreTl4RHM/view

This will be a component only available for Hospitalet (we might generalize it un the future and add it to decidim).

This component will add a feature to processes so that citizens can answer a predefined survey (no surveys can be created at the admin).

Take a look at other components in order to get started.

##### Properties

- Scope (mandatory, select, use scopes to populate the options)
- Categories (mandatory, multiple checkboxes up to 4 options, use categories to populate the options)
- Other priorities (optional, textarea)
- Future ideas (optional, textarea)
- Gender (optional, select, female, male , other)
- Age group (optional, select, 18-24; 25-34;35-44,45-54;55+)
- Zip code (optional, numeric, 5 digits)
- Living at scope? (optional, radio button, yes/no)
- Working at scope? (optional, radio button, yes/no)
- City (optional, select, Hospitalet de Llobregat, Other)
- Name (optional, text)
- Phone (optional, text)

##### Considerations

* Only one submitted survey per user
* Only signed-up users can submit a survey
* After submitting the survey, users should be encouraged to send a proposal
* We don't need an admin interface, it will be specified in another PR.
* The survey results aren't publicly available
* Users can't change their survey submission

##### Implementation details

* Try to KISS but keeping in mind that we'll later generalise this. Maybe we can save all the answers in a single JSON column.

#### :clipboard: Related documentation

Further fields description:
https://docs.google.com/document/d/19enaJK750wC8VIijhoc9tsdJX32fCdNp1_EpGY0X-K0/edit

Example form to use as styles:
http://decidim-design.herokuapp.com/new-proposal/

#### :dart: Acceptance criteria

As a user I can send one submission of the survey. 

#### :pushpin: Related issues

- #11

### PR DESCRIPTION
**Pull request by mrcasals**
#### :tophat: What? Why?
This PR adds surveys for users.

#### :pushpin: Related Issues
- Fixes #13

#### :clipboard: Subtasks
- [x] Make submit save the survey results
- [x] Do not let the user send more than one survey per scope/feature combination
- [x] Add index to not allow this 👆 
- [x] Specs
- [x] I18n
- [x] localize genders
- [x] extract age groups
- [x] correctly set the possible towns
- [x] Extract form collection values in helpers
- [x] Ensure no more than 4 categories can be selected (test it)
- [x] add missing specs for CreatesurveyResult

### :camera: Screenshots (optional)
*None*

### CODE REVIEW

**Comment by mrcasals** [comment]
Tests are green on my machine, I'll need help from @josepjaume in order to make then run on Travis

File: engines/decidim_hospitalet-surveys/Gemfile [hunk 1]
```diff
@@ -0,0 +1,4 @@
+source "https://rubygems.org"
+
+gemspec
+gem "decidim-dev", github: "AjuntamentdeBarcelona/decidim"
```
Comment (oriolgual) [inline engines/decidim_hospitalet-surveys/Gemfile]: Maybe we should move this inside the gemspec as a dev dependency?

File: engines/decidim_hospitalet-surveys/README.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
+# Decidim::Proposals
```
Comment (oriolgual) [inline engines/decidim_hospitalet-surveys/README.md]: Fix readme

File: engines/decidim_hospitalet-surveys/app/controllers/decidim_hospitalet/surveys/survey_results_controller.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+          on(:invalid) do
+            flash.now[:alert] = I18n.t("surveys.create.error", scope: "decidim_hospitalet")
+            render :new
+          end
+        end
+      end
+
+      def available_scopes
+        @available_scopes ||= current_organization.scopes.reject { |scope| answered_surveys.pluck(:decidim_scope_id).include?(scope.id) }
```
Comment (oriolgual) [inline engines/decidim_hospitalet-surveys/app/controllers/decidim_hospitalet/surveys/survey_results_controller.rb]: Can you change this to do it with SQL? Something like `current_organization.scopes.not.where(id: answered_surveys.pluck(:decidim_scope_id))`

File: engines/decidim_hospitalet-surveys/app/controllers/decidim_hospitalet/surveys/survey_results_controller.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+      end
+
+      private
+
+      def answered_surveys
```
Comment (oriolgual) [inline engines/decidim_hospitalet-surveys/app/controllers/decidim_hospitalet/surveys/survey_results_controller.rb]: Maybe we can move this as a method or association to the user? Although I'm not sure how to extend `Decidim::User` /cc @josepjaume

File: engines/decidim_hospitalet-surveys/app/forms/decidim_hospitalet/surveys/survey_form.rb [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+      validates :categories, length: { minimum: 1, maximum: 4 }
+      validates :age_group, inclusion: { in: SurveyResult::AGE_GROUPS }, allow_blank: true
+      validates :gender, inclusion: { in: SurveyResult::GENDERS }, allow_blank: true
+      validates :city, inclusion: { in: Towns::TOWNS.keys }, allow_blank: true
+
+      # Finds the Categories from the category_id.
+      #
+      # Returns a Decidim::Category
+      def categories
+        @category ||= feature.categories.where(id: categories_ids)
```
Comment (oriolgual) [inline engines/decidim_hospitalet-surveys/app/forms/decidim_hospitalet/surveys/survey_form.rb]: s/category/categories

File: engines/decidim_hospitalet-surveys/app/models/decidim_hospitalet/surveys/survey_result.rb [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        return unless user
+        errors.add(:user, :invalid) unless Decidim::User.where(decidim_organization_id: feature.organization.id, id: user.id).exists?
+      end
+
+      def scope_belongs_to_organization
+        return unless scope
+        errors.add(:scope, :invalid) unless feature.scopes.where(id: scope.id).exists?
+      end
+
+      def survey_does_not_exist
```
Comment (oriolgual) [inline engines/decidim_hospitalet-surveys/app/models/decidim_hospitalet/surveys/survey_result.rb]: `validates :scope, uniqueness: { scope: [:feature, :user] }` ?

File: engines/decidim_hospitalet-surveys/app/models/decidim_hospitalet/surveys/survey_result.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        "50-64",
+        "65+"
+      ].freeze
+
+      belongs_to :feature, foreign_key: "decidim_feature_id", class_name: Decidim::Feature
+      belongs_to :user, foreign_key: "decidim_user_id", class_name: Decidim::User
+      belongs_to :scope, foreign_key: "decidim_scope_id", class_name: Decidim::Scope
+      has_one :organization, through: :feature
+
+      validate :user_belongs_to_organization
```
Comment (oriolgual) [inline engines/decidim_hospitalet-surveys/app/models/decidim_hospitalet/surveys/survey_result.rb]: Add missing validates presence

File: engines/decidim_hospitalet-surveys/spec/commands/decidim/proposals/create_proposal_spec.rb [hunk 8]
```diff
@@ -0,0 +1,58 @@
+# frozen_string_literal: true
+require "spec_helper"
+
+module Decidim
+  module Proposals
+    describe CreateProposal do
```
Comment (oriolgual) [inline engines/decidim_hospitalet-surveys/spec/commands/decidim/proposals/create_proposal_spec.rb]: Wrong spec file??

File: engines/decidim_hospitalet-surveys/spec/features/surveys_spec.rb [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    end
+  end
+
+  context "when the user is logged in" do
+    before do
+      login_as user, scope: :user
+    end
+
+    context "when the user has not submitted the survey" do
+      it "creates a new proposal" do
```
Comment (oriolgual) [inline engines/decidim_hospitalet-surveys/spec/features/surveys_spec.rb]: s/proposal/survey result

File: engines/decidim_hospitalet-surveys/spec/shared/admin_shared_context.rb [hunk 10]
```diff
@@ -0,0 +1,11 @@
+RSpec.shared_context "admin" do
```
Comment (oriolgual) [inline engines/decidim_hospitalet-surveys/spec/shared/admin_shared_context.rb]: Not used

File: engines/decidim_hospitalet-surveys/spec/shared/manage_proposals_examples.rb [hunk 11]
```diff
@@ -0,0 +1,40 @@
+# -*- coding: utf-8 -*-
```
Comment (oriolgual) [inline engines/decidim_hospitalet-surveys/spec/shared/manage_proposals_examples.rb]: Not used

**Pull request merged**